purchases/src/main/kotlin/com/revenuecat/purchases/SyncPurchasesHelper.kt

```diff
@@ -0,0 +1,100 @@
+
+        val appUserID = identityManager.currentAppUserID
+
+        billing.queryAllPurchases(
+            appUserID,
+            onReceivePurchaseHistory = { allPurchases ->
+                if (allPurchases.isNotEmpty()) {
+                    val lastPurchase = allPurchases.last()
+                    val errors: MutableList<PurchasesError> = mutableListOf()
+                    fun handleLastPurchase(currentPurchase: StoreTransaction, lastPurchase: StoreTransaction) {
+                        if (currentPurchase == lastPurchase) {
+                            if (errors.isEmpty()) {
+                                debugLog(PurchaseStrings.SYNCED_PURCHASES_SUCCESSFULLY)
+                                retrieveCustomerInfo(appUserID, onSuccess, onError)
+                            } else {
+                                errorLog(PurchaseStrings.SYNCING_PURCHASES_ERROR.format(errors))
+                                onError(errors.first())
+                            }
+                        }
+                    }
+                    allPurchases.forEach { purchase ->
+                        val productInfo = ReceiptInfo(productIDs = purchase.productIds)
+                        postReceiptHelper.postTokenWithoutConsuming(
+                            purchase.purchaseToken,
+                            purchase.storeUserID,
+                            productInfo,
+                            isRestore,
+                            appUserID,
+                            purchase.marketplace,
+                            {
+                                log(LogIntent.PURCHASE, PurchaseStrings.PURCHASE_SYNCED.format(purchase))
+                                handleLastPurchase(purchase, lastPurchase)
+                            },
+                            { error ->
+                                log(
+                                    LogIntent.RC_ERROR, PurchaseStrings.SYNCING_PURCHASES_ERROR_DETAILS
+                                        .format(purchase, error)
+                                )
+                                errors.add(error)
+                                handleLastPurchase(purchase, lastPurchase)
+                            }
+                        )
+                    }
+                } else {
+                    retrieveCustomerInfo(appUserID, onSuccess, onError)
+                }
+            },
+            onReceivePurchaseHistoryError = {
+                log(LogIntent.RC_ERROR, PurchaseStrings.SYNCING_PURCHASES_ERROR.format(it))
+                onError(it)
+            }
+        )
+    }
+
+    private fun retrieveCustomerInfo(
+        appUserID: String,
+        onSuccess: (CustomerInfo) -> Unit,
+        onError: (PurchasesError) -> Unit
+    ) {
+        customerInfoHelper.retrieveCustomerInfo(
+            appUserID,
+            CacheFetchPolicy.CACHED_OR_FETCHED,
+            appInBackground = false,
```

Comment:
Could this `appInBackground = false` be problematic if someone is calling `syncPurchases` from the background?

Reply:
This is used to detect whether the cached customer info is stale or not. If it's in the background, it won't be considered stale until it's 25h old. So potentially, this being hardcoded to false can cause an increase in requests to the get customer info endpoint. 

So I thought that considering the use cases for `syncPurchases`, users shouldn't really call it with the app in the background... But I think it's better to be safe than sorry, so I can pass that as a parameter to avoid that change.